### PR TITLE
Added Outlook Plugin

### PR DESCRIPTION
**Pull request by virlatinus**
| Q  | A
| --- | ---
| Bug fix? | No
| New feature? | Yes
| Related user documentation PR URL | [https://github.com/mautic/documentation/pull/78](https://github.com/mautic/documentation/pull/78)
| Related developer documentation PR URL | None
| Issues addressed (#s or URLs) | None
| BC breaks? | None
| Deprecations? | None

#### Description:
This plugin allows for the Outlook Add-In to keep track of emails sent to leads.

The Add-In source code can be found at [https://github.com/virlatinus/MauticOutlookPlugin/](https://github.com/virlatinus/MauticOutlookPlugin/)

#### Steps to test this PR:
1. Install the Mautic plugin as usual. It will appear on the plugins page in Mautic.
image

2. Click on the Outlook plugin button and enter a secret or key to validate the Outlook Add-In
image

3. Run the [Mautic Outlook Add-In Installer](https://github.com/mautic/mautic/files/402928/Mautic.Outlook.Plugin.Setup.zip) on a Windows machine with Outlook 2016

4. On the Outlook 2016 Options window, select Add-Ins and click on the Add-In Options button
image

5. Type in the URL of the Mautic application, the same secret you used on the Mautic plugin page and click OK
image

6. To track an email sent to a lead, click the Track Email button on the New Email window
image

7. This will append a tracking GIF to the email with the following syntax:  [[MAUTIC_URL]]/index.php/outlook/tracking.gif?d=H4sIAAAAAAAEAIVRTW%2FCMAz9LRyCtgNVlFBpHHroWsRuk8ak7RpaUzqaGCUp0H8%2Fpy0TH4dJUZy892w921uLOvkCa8BGK2WLWi2dt6pUbM7PYPEcFainoFXdJKdBVvUy4quA9rxrNz9Q%2BCQ16HdgmeAenKewpfIU3lvfIO6nGyy75HNXO8LQAN3984R2X5tqMpkwnjOejrfg19%2FBJIHBJsskS3M1MOvOedChUA5HaPBAsp54a7UyBH%2BAw9YWECRrsMc6PHvFd2iR0NfW1QbcjUDwMjhctYqqq0YxkQU6SqMhNxi85GeoD8p0134zaBom%2By4ezlPMxTPFeCH5TLzI%2BdgizeEu5aIUQixmIubjSG5WAY8bC8kyC%2FvxSBX%2Flcvl3bT%2Fvr8k1oBgIQIAAA%3D%3D&sig=cf078d5b

8. The Mautic plugin will then validate the information using the secret to compare signatures and then attach that email to the contact’s profile as part of their activity history. If the lead or leads don't exist, they are created automatically
image
image

### CODE REVIEW

**Comment by alanhartless** [comment]
Hi @virlatinus. Thanks for this and welcome! This is pretty darn cool!

Just a few thoughts: 
1. We need to do some kind of sanitization on the $query Mautic side. The `$mailer` doesn't do this since it's use has historically been behind an authorized session. But for this case, it's a publicly accessible URL that can be used to inject just about anything into a contact's timeline that could then be used to possibly execute rogue script when viewing the email's contents by the Mautic user. So a hacker could submit a known Mautic form with an email. Now there's a contact for that email in Mautic. Then they could hit up this new endpoint with that email injecting rogue script as the body. When Mautic's user checks out that custom email, the rogue script is executed. 

Sanitizing the subject/body Mautic's side will help prevent that. A better option may be to do a simple signature validation. Have the Mautic plugin generate an random key that can be inserted into Outlook plugin. Create a signature of the query data against that key and append to the query param as a signature of the request. (Assuming this is possible by the outlook plugin - no clue even what language that uses). Then on Mautic's PublicController.php side, take the query and use the same process to generate a signature with the key to see if it matches the signature of the query request. This will at least give some basic protection against just anyone trying to inject html. 

2. Lines 80 and 91 of PublicController.php maybe aren't needed? `BuilderTokenHelper::replaceVisualPlaceholdersWithTokens` used to be used with ckeditor when building an email. In fact, since ckeditor was replaced with froala, I think that function is now obsolete and we just need to remove it - (@escopecz, can you verify that?). But I'm assuming any email sent through Outlook won't have any tokens to be replaced by Mautic since the send isn't actually going through Mautic to begin with.

Line 91 shouldn't be needed unless we have a bug that's not sanitizing emojis in the subject when creating the email copy for the stat. 

3. It would be awesome to create the lead(s) if not found. Although this provides another public means to manipulate data without authorization or approval unless  the signature idea of # 1 is implemented.

I look forward to seeing more from you! :-D

**Comment by virlatinus** [comment]
Hi @alanhartless.

Thanks for the great feedback! I had thought at first of adding a field to the plugin options window for storing a secret or key that the Outlook Add-In could use to authenticate.

I already made the changes you suggested and I'm going to submit the PR with those changes.

Best regards,

**Comment by dbhurley** [comment]
+1 excellent work!

**Comment by escopecz** [comment]
2) yes, Alan, that's true.

I don't have a Windows machine to test the full functionality, but this PR itself doesn't break anything in Mautic itself so +1

**Comment by escopecz** [comment]
@virlatinus there is a new rule to document the new features before merge to keep the documentation up to date. Your PR description is basically the documentation. Could you make a PR out of it to https://github.com/mautic/documentation ?

**Comment by TomRoethlisberger** [comment]
+1 on awesomeness

**Comment by Arjoni** [comment]
Can something like this be done for thunderbird? Thunderbird is cross platform and can be used in all OS-s.

File: critical_migrations.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+["20160225000000"]
```
Comment (alanhartless) [inline critical_migrations.txt]: This shouldn't be here - maybe it was committed on accident from an upgrade on the git repo?

File: plugins/MauticOutlookBundle/.gitignore [hunk 2]
```diff
@@ -0,0 +1,49 @@
+# Windows image file caches
```
Comment (alanhartless) [inline plugins/MauticOutlookBundle/.gitignore]: Is this file needed?

File: plugins/MauticOutlookBundle/Config/config.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    'routes'   => array(
+        'public' => [
+            'mautic_outlook_tracker'             => [
+                'path'       => '/outlook/tracking.gif',
+                'controller' => 'MauticOutlookBundle:Public:trackingImage'
+            ]
+        ]
+    ),
+
+    'services'    => array(
```
Comment (alanhartless) [inline plugins/MauticOutlookBundle/Config/config.php]: Are these used? I'm not seeing these actually used anywhere. This `MauticPlugin\MauticOutlookBundle\EventListener\FormSubscriber`,  `MauticPlugin\MauticOutlookBundle\Form\Type\FormFieldsType` seem to be registering a post form submit action that will show up when building a form but they don't seem to be doing anything. Let's remove them.

File: plugins/MauticOutlookBundle/README.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Outlook Plugin for Mautic
```
Comment (alanhartless) [inline plugins/MauticOutlookBundle/README.md]: Let's remove this from here and make sure it's added to the documentation repo if not already.

File: plugins/MauticOutlookBundle/Translations/en_US/messages.ini [hunk 5]
```diff
@@ -0,0 +1,3 @@
+mautic.integration.outlook.secret="Outlook Add-In Secret"
+mautic.plugin.actions.form_to_outlook="Push to Outlook"
```
Comment (alanhartless) [inline plugins/MauticOutlookBundle/Translations/en_US/messages.ini]: Not needed if the form submit action is removed.

File: plugins/MauticOutlookBundle/Controller/PublicController.php [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    /**
+     * @param $lead
+     * @param $email
+     * @param $query
+     * @param $idHash
+     */
+    public function addStat($lead, $email, $query, $idHash)
+    {
+        if ($lead !== null) {
+            $mailer = $this->factory->getMailer();
```
Comment (alanhartless) [inline plugins/MauticOutlookBundle/Controller/PublicController.php]: Use `$this->get('mautic.helper.mailer')->getMailer()` here instead of `$this->factory->getMailer()`

File: plugins/MauticOutlookBundle/EventListener/FormSubscriber.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
+<?php
```
Comment (alanhartless) [inline plugins/MauticOutlookBundle/EventListener/FormSubscriber.php]: Let's delete this file - it doesn't seem to be used.

File: plugins/MauticOutlookBundle/Form/Type/FormFieldsType.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
+<?php
```
Comment (alanhartless) [inline plugins/MauticOutlookBundle/Form/Type/FormFieldsType.php]: Let's delete this file - it doesn't seem to be used.

**Comment by alanhartless** [comment]
I don't have outlook on a windows anywhere but code looks good.

**Comment by arthurscholten** [comment]
I use outlook 2016 local install, hosted at Microsoft. Plus the cloud free version of Mautic. 
I can see that the pixel is placed in an email. After opening the email I can see the trackingpixel link shows-up under reports, pagehits. But the emailaddress is not added to Mautic so I can only see the IP address. In other words. In Mautic I can see that an email was opened, but not which email and by whom. Next the visitor stays anonimous when he visits websites. Is this by design or not working. Furthermore. The URL does not contain UTM.
